warehouse/redshift.go

```diff
@@ -149,23 +149,52 @@ func (rs *Redshift) LoadToWarehouse(file string, _ ...fullstory.ExportMeta) erro
 	}
 	defer rs.conn.Close()
 
-	if !rs.DoesTableExist(rs.conf.Redshift.ExportTable) {
-		if err := rs.CreateExportTable(); err != nil {
-			return err
-		}
+	if err = rs.EnsureCorrectExportTable(); err != nil {
+		return err
 	}
 
-	if err := rs.CopyInData(s3obj); err != nil {
+	schemaHeaders := rs.getSchemaHeaders(rs.exportSchema)
+	if err = rs.CopyInData(s3obj, schemaHeaders); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (rs *Redshift) CopyInData(s3file string) error {
+func (rs *Redshift) EnsureCorrectExportTable() error {
+	var err error
+	if rs.DoesTableExist(rs.conf.Redshift.ExportTable) {
+		// make sure all the columns in the csv export exist in the Export table
+		exportTableColumns := rs.getTableColumns(rs.conf.Redshift.ExportTable)
+		missingFields := rs.getMissingFields(rs.exportSchema, exportTableColumns)
+
+		// If some fields are missing from the fsexport table, either we added new fields
+		// or existing expected columns were deleted by the user we add the relevant columns.
+		// Alter the table and add the missing columns.
+		if len(missingFields) > 0 {
+			log.Printf("Found %d missing fields. Adding columns for these fields.", len(missingFields))
+			alterTableStmts := rs.getAlterTableStatements(missingFields, rs.conf.Redshift.ExportTable)
```

Comment:
nitty nit: I would calculate each alter statement separately, as you need it. Feel free to ignore.